hnix-store-core/src/System/Nix/Hash.hs

```diff
@@ -0,0 +1,105 @@
+{-|
+Description : Nix-style hashes (truncated sha256)
+Maintainer  : Shea Levy <shea@shealevy.com>
+-}
+
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module System.Nix.Hash (
+    StorePathHash (..)
+
+  -- * Introduce hashes for the store
+  , hash
+  , hashlazy
+  , fromBase64Digest
+
+  -- * Eliminate hashes for the store
+  , hashToBase32
+
+  -- * cryptohash-sha256 style incremental hash building
+  , init
+  , update
+  , finalize
+
+  -- * Internal
+  , truncate52
+  , toNixBase32
+
+  ) where
+
+import qualified Crypto.Hash.SHA256           as SHA
+import           Data.Bits
+import qualified Data.ByteString              as BS
+import qualified Data.ByteString.Lazy         as BSL
+import qualified Data.ByteString.Lazy.Builder as BSL
+import qualified Data.Hashable                as Hashable
+import           Data.Semigroup               ((<>))
+import           Data.Word
+import qualified Data.Vector.Unboxed          as UV
+import           Prelude                      hiding (init)
+
+
+newtype StorePathHash = StorePathHash { getTruncatedHash :: BS.ByteString }
+  deriving (Eq, Hashable.Hashable, Ord, Show)
+
+
+init :: SHA.Ctx
+init = SHA.init
+
+
+update :: SHA.Ctx -> BS.ByteString -> SHA.Ctx
+update = SHA.update
+
+
+finalize :: SHA.Ctx -> StorePathHash
+finalize ctx = StorePathHash . truncate52 $ SHA.finalize ctx
+
+
+hash :: BS.ByteString -> StorePathHash
+hash bs = StorePathHash . truncate52 $ SHA.hash bs
+
+
+hashlazy :: BSL.ByteString -> StorePathHash
+hashlazy bs = StorePathHash . truncate52 $ SHA.hashlazy bs
+
+
+-- | Import and validate a store path hash
+fromBase64Digest :: BS.ByteString -> Maybe StorePathHash
+fromBase64Digest = validateRawDigest . StorePathHash
+  where validateRawDigest = Just
+  -- TODO: What should we check for? Only valid base64 chars?
```

Comment:
base64? Do you mean base32? Anyway, we should check that it's of the right length and only has the right charcters.

Wait, why are we putting either of these things into a StorePathHash? That's just a 160 bit bytestring, right? No characters of any kind.